java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -79,53 +79,22 @@ class EntityHandleManager(
         storageKey: StorageKey,
         ttl: Ttl = Ttl.Infinite,
         idGenerator: Id.Generator = Id.Generator.newSession()
-    ): BaseHandleAdapter {
-        val name = idGenerator.newChildId(
-            idGenerator.newChildId(arcId.toArcId(), hostId),
-            baseName
-        ).toString()
-
-        val entityPrep = EntityPreparer<T>(
-            name,
-            idGenerator,
-            entitySpec.SCHEMA,
-            ttl,
-            time
-        )
-
-        val store = stores.get(
-            SingletonStoreOptions<RawEntity>(
-                storageKey = storageKey,
-                type = SingletonType(EntityType(entitySpec.SCHEMA)),
-                mode = ReferenceMode
-            )
-        ).activate(activationFactory)
-
-        val storageProxy = singletonStorageProxies.getOrPut(storageKey) {
-            SingletonProxy(store, CrdtSingleton())
-        }
-
-        return when (mode) {
-            HandleMode.Read -> ReadSingletonHandleAdapter(
-                name = name,
-                entitySpec = entitySpec,
-                storageProxy = storageProxy,
-                dereferencerFactory = dereferencerFactory
-            )
-            HandleMode.Write -> WriteSingletonHandleAdapter<T>(
-                name = name,
-                storageProxy = storageProxy,
-                entityPreparer = entityPrep
-            )
-            HandleMode.ReadWrite -> ReadWriteSingletonHandleAdapter(
-                name = name,
+    ): Handle =
+        idGenerator.handleName(baseName).let { name ->
+            SingletonHandle(
+                name = idGenerator.handleName(baseName),
                 entitySpec = entitySpec,
-                storageProxy = storageProxy,
-                entityPreparer = entityPrep,
+                storageProxy = singletonStoreProxy(storageKey, entitySpec),
+                entityPreparer = EntityPreparer(name, idGenerator, entitySpec.SCHEMA, ttl, time),
                 dereferencerFactory = dereferencerFactory
             )
+        }.let { handle ->
+            return when (mode) {
```

Comment:
Nice. Although we've come full circle back to facades again, this trick is a lot similar the original one I was doing in BaseParticle and my original PR.